### ISSUE TITLE
[REVIEW]: LFSpy: A Python Implementation of Local Feature Selection for Data Classification with scikit-learn Compatibility

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#1958]
**Submitting author:** @kiretd (<a href="http://orcid.org/0000-0003-4849-732X">Kiret Dhindsa</a>)
**Repository:** <a href="https://github.com/McMasterRS/LFSpy" target ="_blank">https://github.com/McMasterRS/LFSpy</a>
**Version:** v1.0.4
**Editor:** @dfm
**Reviewers:** @effigies, @sauln
**Archive:** <a href="https://doi.org/10.5281/zenodo.3813708" target="_blank">10.5281/zenodo.3813708</a>

## Status

[![status](https://joss.theoj.org/papers/e8b1f0500a1397299d7f2ab30bcd0afb/status.svg)](https://joss.theoj.org/papers/e8b1f0500a1397299d7f2ab30bcd0afb)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/e8b1f0500a1397299d7f2ab30bcd0afb"><img src="https://joss.theoj.org/papers/e8b1f0500a1397299d7f2ab30bcd0afb/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/e8b1f0500a1397299d7f2ab30bcd0afb/status.svg)](https://joss.theoj.org/papers/e8b1f0500a1397299d7f2ab30bcd0afb)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/McMasterRS/LFSpy" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@effigies & @mnarayan, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @dfm know.

✨ **Please try and complete your review in the next two weeks** ✨ 

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/McMasterRS/LFSpy">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@kiretd) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/McMasterRS/LFSpy">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@kiretd) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @effigies, @mnarayan it looks like you're currently assigned to review this paper :tada:.

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
```
Attempting to check references...
```

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
```Reference check summary:

OK DOIs

- 10.1109/TPAMI.2015.2478471 is OK
- 10.1109/TNNLS.2017.2676101 is OK
- 10.1109/JBHI.2018.2877738 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01958/joss.01958/10.21105.joss.01958.pdf)

**Comment by effigies** [comment]
I've started on this, but may not finish until after the holidays. Apologies if you need a quicker review.

**Comment by dfm** [comment]
@effigies: No worries and thanks for letting us know! I'll remind you in the new year if we don't hear from you.

**Comment by kiretd** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01958/joss.01958/10.21105.joss.01958.pdf)

**Comment by dfm** [comment]
@effigies, @mnarayan: Happy new year! I'm checking in to remind you about this review and to see if you have any questions going through the review checklist. Please let me know if there are any issues. Thanks!

**Comment by kiretd** [comment]
Hi @dfm, Just wondering if there's anything we should be doing on our end?

**Comment by effigies** [comment]
Ah, sorry. I'll try to have a look on Friday.

**Comment by dfm** [comment]
Thanks @effigies!

@mnarayan: Any updates from your end? Thanks!

**Comment by effigies** [comment]
## Review

LFSpy is a Python reimplementation of [LFS](https://github.com/armanfn/LFS), originally written by Narges Armanfard (@armanfn).

Much of the requested review is straightforward box-checking, so I'm restricting these comments to those where I have questions.

> ### General checks
>
> * [ ]  **Contribution and authorship:** Has the submitting author (@kiretd) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

@kiretd wrote the paper and also contributed tests. According to the commit log, the contributors to the repository were:

```
    28  Oliver Cook <cooko@mcmaster.ca>
    23  Areeb Khawaja <khawaa1@mcmaster.ca>
    16  Kiret Dhindsa <kiretd@gmail.com>
    10  Tom Mudway <mudwayt@mcmaster.ca>
```

Areeb Khawaja appears to be missing from the author list, and Ron Harwood and Ranil Sonnadara have not contributed. Ranil appears to be a PI and Ron a staff member. I'll simply link to the [Authorship guidelines](https://joss.readthedocs.io/en/latest/submitting.html#authorship) and would appreciate an affirmation that Ron and Ranil provided substantial contribution to the effort, and that @khawaa1 has elected not to be an author on this submission.

> ### Documentation

Documentation currently has the form of a [README.md](https://github.com/McMasterRS/LFSpy/blob/master/README.md). Although there are extensive comments in the code, there is not rendered documentation anywhere.

> * [ ]  **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?

This is currently missing

> * [x]  **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.

This is there, although the installation requirements currently include `pytest`. The convention typically is to set installation requirements only to those needed to use the software, and to include tests and optional functionality as `extras_require` arguments to the `setuptools.setup` function. I will submit a pull request to the main repository to use this. (Authors are not obligated to accept this. Some projects prefer to treat tests as first-class requirements.)

> * [ ]  **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).

The [example usage](https://github.com/McMasterRS/LFSpy#usage) is not directly runnable. From the perspective of users of [scikit-learn](https://scikit-learn.org), it is very likely enough, but fully runnable examples would be preferable, IMO. The [test_lfs.py](https://github.com/McMasterRS/LFSpy/blob/master/LFSpy/tests/test_lfs.py) module fetches two datasets. One of these could be used to flesh out the example usage.

> * [ ]  **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?

No. Examples of standard scikit-learn API documentation strings can be found [here](https://github.com/scikit-learn/scikit-learn/blob/b194674c4/sklearn/linear_model/_base.py#L375).

For these to be readable (as in [sklearn.linear_model.LinearRegression](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LinearRegression.html#sklearn.linear_model.LinearRegression)), the documentation would need to be rendered.

> * [x]  **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?

Yes, however, tests fail. Opened https://github.com/McMasterRS/LFSpy/issues/3.

> * [ ]  **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

This is missing.

> ### Software paper

> * [ ]  **State of the field:** Do the authors describe how this software compares to other commonly-used packages?

This currently feels implicit, and explicit comparisons to other methods would be helpful. I think comparisons to other classifiers within scikit-learn would serve equally as well as independent packages, if that's easier.

I suspect the authors are more familiar with the current state of the art in machine learning than I am, but if it's useful: My impression is that random forests (or other decision tree ensembles) will do a similar task, splitting the data into subspaces and selecting the most effective dimensions to distinguish classes.

**Comment by dfm** [comment]
@mnarayan: Can you please let us know when you'll have a chance to look at this submission? Thanks!

**Comment by kiretd** [comment]
Hi @effigies, thanks for your review!

> > ### General checks
> > 
> > * [ ]   **Contribution and authorship:** Has the submitting author (@kiretd) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
> 
> @kiretd wrote the paper and also contributed tests. According to the commit log, the contributors to the repository were:

Sorry that the contributions are a bit difficult to track. Most of the work on LFSpy was done prior to putting this up on github, so the github commits don’t reflect everyone’s contributions. In addition, a good chunk of the testing we do is done locally as a group before someone pushes it to git. If you have further questions about who contributed what, please feel free to ask. 

> Areeb Khawaja appears to be missing from the author list, and Ron Harwood and Ranil Sonnadara have not contributed. Ranil appears to be a PI and Ron a staff member. I'll simply link to the [Authorship guidelines](https://joss.readthedocs.io/en/latest/submitting.html#authorship) and would appreciate an affirmation that Ron and Ranil provided substantial contribution to the effort, and that @khawaa1 has elected not to be an author on this submission.

Both Ron Harwood and Ranil Sonnadara were responsible for direction of the project, including leading the software planning, requirements gathering, and some of the user testing necessary for development of the final project. 
I’ve also added Areeb to the authorship now that he has an ORCID. 


> > ### Documentation
> 
> Documentation currently has the form of a [README.md](https://github.com/McMasterRS/LFSpy/blob/master/README.md). Although there are extensive comments in the code, there is not rendered documentation anywhere.

We now have a separate documentation page at https://lfspy.readthedocs.io/en/latest/
It’s not 100% complete yet, though it’s not too far off, and we’ll be keeping this up to date over the lifetime of this project. Please take a look and let us know if there are any pieces missing.

> > * [ ]   **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
> 
> This is currently missing

We’ve added a Statement of Need to the docs page and the README on github.

> > * [x]   **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
> This is there, although the installation requirements currently include `pytest`. The convention typically is to set installation requirements only to those needed to use the software, and to include tests and optional functionality as `extras_require` arguments to the `setuptools.setup` function. I will submit a pull request to the main repository to use this. (Authors are not obligated to accept this. Some projects prefer to treat tests as first-class requirements.)

This has been updated per your (generous) recommendations on the issues thread on Github. 

> > * [ ]   **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
> 
> The [example usage](https://github.com/McMasterRS/LFSpy#usage) is not directly runnable. From the perspective of users of [scikit-learn](https://scikit-learn.org), it is very likely enough, but fully runnable examples would be preferable, IMO. The [test_lfs.py](https://github.com/McMasterRS/LFSpy/blob/master/LFSpy/tests/test_lfs.py) module fetches two datasets. One of these could be used to flesh out the example usage.

Great suggestion! I’ve added an example to the README and docs, and an example.py to the repository.

> > * [ ]   **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
> 
> No. Examples of standard scikit-learn API documentation strings can be found [here](https://github.com/scikit-learn/scikit-learn/blob/b194674c4/sklearn/linear_model/_base.py#L375).
> 
> For these to be readable (as in [sklearn.linear_model.LinearRegression](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LinearRegression.html#sklearn.linear_model.LinearRegression)), the documentation would need to be rendered.

Another great suggestion! We’ve added this here: https://lfspy.readthedocs.io/en/latest/Functionality.html

> > * [x]   **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
> 
> Yes, however, tests fail. Opened [McMasterRS/LFSpy#3](https://github.com/McMasterRS/LFSpy/issues/3).

Thanks for finding this. Apparently an update broke the test, but it should be fixed now (see response on the repo).

> > * [ ]   **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support
> 
> This is missing.

We now have a contribution guidelines page on our docs: https://lfspy.readthedocs.io/en/latest/Community%20Guidelines.html

> > ### Software paper
> 
> > * [ ]   **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
> 
> This currently feels implicit, and explicit comparisons to other methods would be helpful. I think comparisons to other classifiers within scikit-learn would serve equally as well as independent packages, if that's easier.
> 
> I suspect the authors are more familiar with the current state of the art in machine learning than I am, but if it's useful: My impression is that random forests (or other decision tree ensembles) will do a similar task, splitting the data into subspaces and selecting the most effective dimensions to distinguish classes.

I’ve added a section on comparison with other standard classifiers, including plots using both sample datasets used in test_lfs.py. You were correct that random forests have some similarities and are a good point of comparison. I’ve also included another standard approach; a linear SVM with univariate feature selection using the F-statistic. These comparisons are in comparisons.py, which has been added to the tests folder.

**Comment by kiretd** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
PDF failed to compile for issue #1958 with the following error: 

 ORCID looks to be the wrong length
/app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-9847f98e9ec6/lib/whedon.rb:153:in `block in check_orcids': Problem with ORCID (00000-0003-4528-9146) for Areeb Khawaja (RuntimeError)
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-9847f98e9ec6/lib/whedon.rb:151:in `each'
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-9847f98e9ec6/lib/whedon.rb:151:in `check_orcids'
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-9847f98e9ec6/lib/whedon.rb:88:in `initialize'
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-9847f98e9ec6/lib/whedon/processor.rb:36:in `new'
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-9847f98e9ec6/lib/whedon/processor.rb:36:in `set_paper'
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-9847f98e9ec6/bin/whedon:55:in `prepare'
	from /app/vendor/bundle/ruby/2.4.0/gems/thor-0.20.3/lib/thor/command.rb:27:in `run'
	from /app/vendor/bundle/ruby/2.4.0/gems/thor-0.20.3/lib/thor/invocation.rb:126:in `invoke_command'
	from /app/vendor/bundle/ruby/2.4.0/gems/thor-0.20.3/lib/thor.rb:387:in `dispatch'
	from /app/vendor/bundle/ruby/2.4.0/gems/thor-0.20.3/lib/thor/base.rb:466:in `start'
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-9847f98e9ec6/bin/whedon:116:in `<top (required)>'
	from /app/vendor/bundle/ruby/2.4.0/bin/whedon:23:in `load'
	from /app/vendor/bundle/ruby/2.4.0/bin/whedon:23:in `<main>'

**Comment by kiretd** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
PDF failed to compile for issue #1958 with the following error: 

 [WARNING] Could not convert image '/tmp/tex2pdf.-f4119cc7dbe480a9/20ced9490086bcac201d809020b064159913517e.shtml': Cannot load file
  Jpeg Invalid marker used
  PNG Invalid PNG file, signature broken
  Bitmap Invalid Bitmap magic identifier
  GIF Invalid Gif signature : 
  
  
  
  
  
  
  HDR Invalid radiance file signature
  Tiff Invalid endian tag value
  TGA Invalid bit depth (104)
[WARNING] Could not convert image '/tmp/tex2pdf.-f4119cc7dbe480a9/1ff300c76fa24ed9469b4aea44b91258d01e7f3d.shtml': Cannot load file
  Jpeg Invalid marker used
  PNG Invalid PNG file, signature broken
  Bitmap Invalid Bitmap magic identifier
  GIF Invalid Gif signature : 
  
  
  
  
  
  
  HDR Invalid radiance file signature
  Tiff Invalid endian tag value
  TGA Invalid bit depth (104)
Error producing PDF.
! LaTeX Error: Cannot determine size of graphic in /tmp/tex2pdf.-f4119cc7dbe480
a9/20ced9490086bcac201d809020b064159913517e.shtml (no BoundingBox).

See the LaTeX manual or LaTeX Companion for explanation.
Type  H <return>  for immediate help.
 ...                                              
                                                  
l.522 ...490086bcac201d809020b064159913517e.shtml}

Looks like we failed to compile the PDF

**Comment by kiretd** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01958/joss.01958/10.21105.joss.01958.pdf)

**Comment by kiretd** [comment]
@whendon generate pdf

**Comment by kiretd** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01958/joss.01958/10.21105.joss.01958.pdf)

**Comment by kiretd** [comment]
@dfm Hey, sorry about this, but I think I need help including figures in the pdf. Using standard markdown syntax (same as the example given on the submission instructions page), whedon spits out the error a few posts above. I switched to html because it was also the only method that allowed me to resize the images, and now the pdf compiles without images. Hopefully I'm just missing something really obvious?

**Comment by dfm** [comment]
@kiretd: I just opened a pull request fixing the syntax.

**Comment by kiretd** [comment]
@dfm Thanks so much. It looks like the images just needed to be in the same folder for whedon to find them?
Also, is there no way to resize the images with markdown syntax? I assume the latex conversion won't work if the images are included using html syntax.

**Comment by kiretd** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01958/joss.01958/10.21105.joss.01958.pdf)

**Comment by kiretd** [comment]
Ok, I think everything should be ready for review. I believe we addressed everything from the last review. Thanks again @dfm and @effigies !

**Comment by dfm** [comment]
Since @mnarayan seems to have lost track of this review, I'm currently looking for another referee and I'll keep you all posted.

**Comment by dfm** [comment]
@effigies: Thanks for your review so far! When you get a chance, please take a look at the changes that @kiretd has made and continue through the checklist. Thanks again!

**Comment by dfm** [comment]
@sauln has agreed to take over for mnarayan. Thanks @sauln!

I'll add you as a reviewer and update the checklist. Please let me know if you have any questions or issues as you review.

**Comment by dfm** [comment]
@whedon remove @mnarayan as reviewer

**Comment by whedon** [comment]
OK, @mnarayan is no longer a reviewer

**Comment by dfm** [comment]
@whedon add @sauln as reviewer

**Comment by whedon** [comment]
OK, @sauln is now a reviewer

**Comment by sauln** [comment]
👋 Looks like @effigies did a great review. I'm hoping to have my review complete by early next week.

**Comment by kiretd** [comment]
@whedon generate pdf

**Comment by kiretd** [comment]
Just small update to add the figure captions

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01958/joss.01958/10.21105.joss.01958.pdf)

**Comment by sauln** [comment]
I'll have to provide a more thorough review near the beginning of next week, but I can supply a few quick notes from pursuing the documentation:
- The docs link in the readme doesn't work. You'll need to use `[lfspy.readthedocs.io](https://lfspy.readthedocs.io)`
- It looks like you manually wrote the Functionality page and I see a lot of good comments in the code. If you reformatted the docstrings to comply with an [autodocs](http://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html) format, they would render with the `__doc__` magic method, and you could use autodocs to render the functionality page.  
- It looks like each section in the [homepage](https://lfspy.readthedocs.io/en/latest/index.html) of the docs is repeated as an individual page.  Is this necessary? 
- The contribution guide says that contributions should follow PEP8, but the codebase looks like it doesn't follow PEP8 itself.  Have you ran a linter or autoformatter over code? I highly recommend [black](https://black.readthedocs.io/en/stable/) but any will do.
- Could you provide more examples with larger data sets? 
- The table of tunable parameters in the docs homepage hasn't rendered correctly.
- The tests appear to be end-to-end tests.  I would like to see more unit tests that focus on individual pieces of logic.

**Comment by dfm** [comment]
Hi team! Just checking in here on the status.

@kiretd: I wanted to double check that you got @sauln's recommendations. Let us know if you have any questions!

@effigies: It seems like there might be some remaining issues for you. Can you let us know if there's more that you would like to see beyond this response: https://github.com/openjournals/joss-reviews/issues/1958#issuecomment-583011174

Thanks!

**Comment by effigies** [comment]
Hi, thanks for the reminder. I'll try to have a look on Friday.

**Comment by effigies** [comment]
@dfm My concerns have been addressed.

I submitted a minor PR (https://github.com/McMasterRS/LFSpy/pull/6) for some cleanups in the documentation, but that should not be considered blocking.

**Comment by kiretd** [comment]
Hi @dfm
Thanks for checking in. Yes, I did get the recommendations by @sauln (thanks for those @sauln). I just moved continents last week, so I haven’t gotten much done on this myself yet. The team has been working on some of the recommendations in the meantime, and hopefully we’ll have an update soon. 
 
@sauln:
Thanks for your suggestions! I’ll respond one by one.

> * The docs link in the readme doesn't work. You'll need to use `[lfspy.readthedocs.io](https://lfspy.readthedocs.io)`

Thanks, looks fixed on our end now.

> * It looks like you manually wrote the Functionality page and I see a lot of good comments in the code. If you reformatted the docstrings to comply with an [autodocs](http://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html) format, they would render with the `__doc__` magic method, and you could use autodocs to render the functionality page.

This sounds really great, and I think we’ll likely start implementing it for our projects going forward. I’m not sure there’s much value in spending the time to redo our comments and functionality page for LFSpy right now, but we’re taking a look at whether it makes sense to switch it over gradually over the next few months or so. 

> * It looks like each section in the [homepage](https://lfspy.readthedocs.io/en/latest/index.html) of the docs is repeated as an individual page.  Is this necessary?

Maybe not necessary, but the way we have our hyperlinked table of contents set up is that each section has a link on the sidebar for easier navigation. 

> * The contribution guide says that contributions should follow PEP8, but the codebase looks like it doesn't follow PEP8 itself.  Have you ran a linter or autoformatter over code? I highly recommend [black](https://black.readthedocs.io/en/stable/) but any will do.

Oliver from our team ran it through Black as you recommended. Pretty cool tool! I’m not an expert on this part, but I believe that it’s not exactly formatted correctly even after using Black. For now, we’ve just removed the PEP8 guideline so there’s no inconsistency.

> * Could you provide more examples with larger data sets?

Do you have any particular datasets in mind? Typically sample datasets used for illustration are small so that the tests and examples can be run relatively quickly. In addition, LFS is particularly well suited for p>>n and small n problems, as per our statement of need, so we think our examples (one custom, the other the standard sk-learn sample dataset used in many of their examples) are rather suitable. 

> * The table of tunable parameters in the docs homepage hasn't rendered correctly.
 
This looks to be fixed on our end. Can you confirm it looks right on your end too?

> * The tests appear to be end-to-end tests.  I would like to see more unit tests that focus on individual pieces of logic.

LFSpy is meant to be run as a single function, usually in a pipeline, but at least as a classifier or a feature selector (using the same function). Do you mean we should publish tests of internal functions as would be used for internal validation?

**Comment by kiretd** [comment]
> I submitted a minor PR ([McMasterRS/LFSpy#6](https://github.com/McMasterRS/LFSpy/pull/6)) for some cleanups in the documentation, but that should not be considered blocking.

Thanks @effigies looks good to me, just waiting for one more person to review in case I missed something. 

@dfm Just want to check in to see where we are with the process. I know I took some time to respond while we made the updates, so hopefully nothing is waiting on me right now.

**Comment by dfm** [comment]
@sauln: Do you feel that @kiretd has addressed some of your concerns? Do you have more feedback?

@kiretd: In response to your question to @sauln:

> Do you mean we should publish tests of internal functions as would be used for internal validation?

I believe that that was what he was saying. In general, [unit tests](https://en.wikipedia.org/wiki/Unit_testing) should test each sub-component of the code, not just the end-to-end result. Unit testing, specifically, is not a *requirement* for JOSS, but it is recommended best practice.

**Comment by kiretd** [comment]
@dfm and @sauln Thanks, yep, unit testing makes sense. I'm going to look into whether we have the resources to add that in now, or if we'll need to add it in gradually. I'm looking into exactly what would be the best way to do it now.

**Comment by arfon** [comment]
Dear authors and reviewers

We wanted to notify you that in light of the current COVID-19 pandemic, JOSS has decided to suspend submission of new manuscripts and to handle existing manuscripts (such as this one) on a "best efforts basis". We understand that you may need to attend to more pressing issues than completing a review or updating a repository in response to a review. If this is the case, a quick note indicating that you need to put a "pause" on your involvement with a review would be appreciated but is not required.

Thanks in advance for your understanding.

_Arfon Smith, Editor in Chief, on behalf of the JOSS editorial team._

**Comment by sauln** [comment]
Sorry for the delay. I was on PTO, and then the whole global pandemic thing.

Many of my suggested edits were taken care of. The most important of those that were not addressed, I addressed myself in a PR:
[#7](https://github.com/McMasterRS/LFSpy/pull/7) establishes proper docstrings so the code is self-documenting, and
[#8](https://github.com/McMasterRS/LFSpy/pull/8) tweaks the `black` parameters so the formatting isn't so strange looking.

I highly suggest breaking the `fit` method into smaller pieces. At its deepest, I count 14 levels of indentation.  That suggests to me the library will be very difficult to maintain moving forward. 


> In addition, LFS is particularly well suited for p>>n and small n problems, as per our statement of need, so we think our examples (one custom, the other the standard sk-learn sample dataset used in many of their examples) are rather suitable.

The iris data set has p=4 and n=150. Maybe the scikit-learn wine dataset would be a richer example with 13 dimensions? Another option would be to add a handful of examples using the [scikit-learn generated datasets API](https://scikit-learn.org/stable/datasets/index.html#generated-datasets). Then you could really show off the strengths.

**Comment by sauln** [comment]
The above are just strong suggestions @dfm.  If my two PRs are merged, I believe my concerns would be sufficiently addressed.

Thank you.

**Comment by dfm** [comment]
Hi all, my turn to apologize for the slow response!
Thanks for those submitting those pull requests, @sauln. They have now been merged, so can you take a last look at the checklist when you have time and let us know if there are any outstanding issues? Thanks!

**Comment by kiretd** [comment]
Hi @dfm and @sauln, thanks for the updates!

We did indeed merge the changes Saul generously provided, and my delay in response is because we're still working on some of the others (though they were listed as optional). I had a draft response, so I'll just put it below so it's clear where we are now. 

> Many of my suggested edits were taken care of. The most important of those that were not addressed, I addressed myself in a PR:
> [#7](https://github.com/McMasterRS/LFSpy/pull/7) establishes proper docstrings so the code is self-documenting, and
> [#8](https://github.com/McMasterRS/LFSpy/pull/8) tweaks the `black` parameters so the formatting isn't so strange looking.

If I remember correctly, I think we tried to do this, but it didn't look quite right. It looks like you fixed the issue with it, so thanks!

> I highly suggest breaking the `fit` method into smaller pieces. At its deepest, I count 14 levels of indentation. That suggests to me the library will be very difficult to maintain moving forward.

This is actually something we've been working on, and updates on this will come. The way it is now is definitely far from best coding practice, and it won't stay that way. It will take us some time though.

> > In addition, LFS is particularly well suited for p>>n and small n problems, as per our statement of need, so we think our examples (one custom, the other the standard sk-learn sample dataset used in many of their examples) are rather suitable.
> 
> The iris data set has p=4 and n=150. Maybe the scikit-learn wine dataset would be a richer example with 13 dimensions? Another option would be to add a handful of examples using the [scikit-learn generated datasets API](https://scikit-learn.org/stable/datasets/index.html#generated-datasets). Then you could really show off the strengths.

That is correct, and all four features are discriminative. We add increasing numbers of non-discriminative features to showcase exactly what LFSpy excels at (p>>n, but most of those p are not actually useful) so we can simulate a noise dataset. I'm now also building a similar example using scikit-learn's sklearn.datasets.make_classification function, which does have some nice options for controlling the properties of generated datasets. It's taking a bit of time, but I'm hoping it'll be added to the paper and the docs by the end of the week.

In the meantime, it would be great to know if there's anything else we should work on prior to publication. Thanks!

**Comment by kiretd** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01958/joss.01958/10.21105.joss.01958.pdf)

**Comment by kiretd** [comment]
@dfm @sauln 
I've updated the paper and the repo with a new experiment that I think will satisfy one of the requests. 
Cheers all, hope you're in good health.

**Comment by sauln** [comment]
Looks great! 👍

**Comment by dfm** [comment]
Thanks @sauln! Can you take a look at the checklist at the top of this issue to see if there are any that still have open questions? Once you're happy, I'll do some last checks and we'll be ready to proceed with acceptance.

**Comment by sauln** [comment]
Yup, sorry! Everything is checked now.

**Comment by dfm** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01958/joss.01958/10.21105.joss.01958.pdf)

**Comment by dfm** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Reference check summary:

OK DOIs

- 10.1109/TPAMI.2015.2478471 is OK
- 10.1109/TNNLS.2017.2676101 is OK
- 10.1109/JBHI.2018.2877738 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by dfm** [comment]
@kiretd: Thanks for your patience! I've just submitted two pull requests with small changes to the paper. Once you've accepted those, take a last proof read, then:

1. Bump the version number and tag a new release of the code
2. Create an archive of the code using Zenodo or something similar
3. Make sure that the metadata (title and author list, for example) of the archive match the JOSS paper (on Zenodo you can edit the metadata after creating the archive)

Once you've completed those steps, please post the new version number and the DOI for the archive to this thread. Thanks!

**Comment by dfm** [comment]
Also: Thanks for your review @sauln! I'm still seeing the "statement of need" checkbox unchecked, but I'll proceed now assuming that you're planning on checking it :)

**Comment by kiretd** [comment]
Thanks @dfm, as well as @effigies and @sauln for all of the help and effort!

1. We just tagged a release of the code. It's at version 1.0 on Github, but version 1.0.4 on Pypi. Do these need to be the same?
2. DOI: 10.5281/zenodo.3813708
3. Looks right to me.

I've made all of the changes I want to the paper. Did I provide you with everything you need?

**Comment by dfm** [comment]
It's good practice to synchronize versions across PyPI and GitHub releases, but it's not absolutely necessary.

That DOI isn't resolving and I can't find the archive on Zenodo manually. Can you share the Zenodo link directly?

**Comment by kiretd** [comment]
Here's the link, is it working for you?
https://zenodo.org/record/3813708#.XrRVK6gzZPY

We're resolving the version number now.

**Comment by effigies** [comment]
DOI resolves for me. I've noticed Zenodo DOIs have a bit of an increased lag before they resolve, lately.

**Comment by dfm** [comment]
Great! Yes - it's resolving now. Please update the metadata in the Zenodo archive so that the title matches the paper: "LFSpy: A Python Implementation of Local Feature Selection for Data Classification with scikit-learn Compatibility".

**Comment by kiretd** [comment]
Oops, didn't realize it applied to the title as well. Should be good now.

**Comment by dfm** [comment]
@whedon set 10.5281/zenodo.3813708 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.3813708" target="_blank">10.5281/zenodo.3813708</a> is the archive.

**Comment by dfm** [comment]
@whedon set v1.0.4 as version

**Comment by whedon** [comment]
OK. v1.0.4 is the version.

**Comment by dfm** [comment]
@whedon accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```
Reference check summary:

OK DOIs

- 10.1109/TPAMI.2015.2478471 is OK
- 10.1109/TNNLS.2017.2676101 is OK
- 10.1109/JBHI.2018.2877738 is OK
- 10.1109/TNSRE.2019.2922553 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
:wave: @openjournals/joss-eics, this paper is ready to be accepted and published.

 Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/1442

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/1442, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by kiretd** [comment]
Nice, it looks good to me, from what I can tell.

**Comment by arfon** [comment]
@whedon accept deposit=true

**Comment by whedon** [comment]
```
Doing it live! Attempting automated processing of paper acceptance...
```

**Comment by whedon** [comment]
🐦🐦🐦 👉 [Tweet for this paper](https://twitter.com/JOSS_TheOJ/status/1259594215929122816) 👈 🐦🐦🐦

**Comment by whedon** [comment]
🚨🚨🚨 **THIS IS NOT A DRILL, YOU HAVE JUST ACCEPTED A PAPER INTO JOSS!** 🚨🚨🚨

 Here's what you must now do:

0. Check final PDF and Crossref metadata that was deposited :point_right: https://github.com/openjournals/joss-papers/pull/1443
1. Wait a couple of minutes to verify that the paper DOI resolves [https://doi.org/10.21105/joss.01958](https://doi.org/10.21105/joss.01958)
2. If everything looks good, then close this review issue.
3. Party like you just published a paper! 🎉🌈🦄💃👻🤘

 Any issues? notify your editorial technical team...

**Comment by arfon** [comment]
@effigies, @sauln - many thanks for your reviews here and to @dfm for editing this submission ✨ 

@kiretd - your paper is now accepted into JOSS :zap::rocket::boom:

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](https://joss.theoj.org/papers/10.21105/joss.01958/status.svg)](https://doi.org/10.21105/joss.01958)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/joss.01958">
  <img src="https://joss.theoj.org/papers/10.21105/joss.01958/status.svg" alt="DOI badge" >
</a>

reStructuredText:
.. image:: https://joss.theoj.org/papers/10.21105/joss.01958/status.svg
   :target: https://doi.org/10.21105/joss.01958
```
This is how it will look in your documentation:

[![DOI](https://joss.theoj.org/papers/10.21105/joss.01958/status.svg)](https://doi.org/10.21105/joss.01958)

**We need your help!**

Journal of Open Source Software is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: https://joss.theoj.org/reviewer-signup.html
- Making a small donation to support our running costs here: https://numfocus.org/donate-to-joss

**Comment by dfm** [comment]
@effigies, @sauln: Also a huge thanks from me for your reviews!

@kiretd: Congrats on your paper!

**Comment by kiretd** [comment]
@dfm, @effigies, @sauln 
Thanks all! Now that it's all done, I just want to say that this is easily the most transparent, innovative, and overall best review phase I've been through with a journal. I really hope this is what peer review evolves into as the years go on. Not just that, but you all were also extremely helpful and educational throughout the process, so really well done.
